File: implementations/java/client/src/main/java/com/theshuai/specusclient/peer/PeerUdpProbeRateLimiter.java
package com.theshuai.specusclient.peer;

import java.net.InetAddress;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

final class PeerUdpProbeRateLimiter {
    private static final long WINDOW_MILLIS = 1_000;
    private static final long SOURCE_TTL_MILLIS = 60_000;
    private static final int GLOBAL_PACKETS_PER_WINDOW = 2_000;
    private static final int SOURCE_PACKETS_PER_WINDOW = 100;
    private static final int MAX_SOURCES = 4_096;

    private final Map<InetAddress, Window> sourceWindows = new LinkedHashMap<>(64, 0.75f, true);
    private long globalWindowStartedMillis;
    private int globalPackets;

    synchronized boolean tryAcquire(InetAddress source, long nowMillis) {
        if (nowMillis - globalWindowStartedMillis >= WINDOW_MILLIS) {
            globalWindowStartedMillis = nowMillis;
            globalPackets = 0;
        }
        if (++globalPackets > GLOBAL_PACKETS_PER_WINDOW || source == null) {
            return false;
        }

        Window window = sourceWindows.get(source);
        if (window == null) {
            evictIfNeeded(nowMillis);
            window = new Window(nowMillis, 0, nowMillis);
            sourceWindows.put(source, window);
        }
        if (nowMillis - window.startedMillis >= WINDOW_MILLIS) {
            window.startedMillis = nowMillis;
            window.packets = 0;
        }
        window.lastSeenMillis = nowMillis;
        return ++window.packets <= SOURCE_PACKETS_PER_WINDOW;
    }

    synchronized void cleanup(long nowMillis) {
        sourceWindows.entrySet().removeIf(entry -> nowMillis - entry.getValue().lastSeenMillis > SOURCE_TTL_MILLIS);
    }

    private void evictIfNeeded(long nowMillis) {
        Iterator<Map.Entry<InetAddress, Window>> iterator = sourceWindows.entrySet().iterator();
        while (iterator.hasNext() && sourceWindows.size() >= MAX_SOURCES) {
            Map.Entry<InetAddress, Window> entry = iterator.next();
            if (nowMillis - entry.getValue().lastSeenMillis > SOURCE_TTL_MILLIS || sourceWindows.size() >= MAX_SOURCES) {
                iterator.remove();
            }
        }
    }

    private static final class Window {
        private long startedMillis;
        private int packets;
        private long lastSeenMillis;

        private Window(long startedMillis, int packets, long lastSeenMillis) {
            this.startedMillis = startedMillis;
            this.packets = packets;
            this.lastSeenMillis = lastSeenMillis;
        }
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerUdpProbeRateLimiter.cs
using System.Net;

namespace Specus.Client.PeerMesh;

/// <summary>Fixed-window probe limiter aligned with the Java client.</summary>
internal sealed class PeerUdpProbeRateLimiter
{
    internal const long WindowMilliseconds = 1_000;
    internal const long SourceTtlMilliseconds = 60_000;
    internal const int GlobalPacketsPerWindow = 2_000;
    internal const int SourcePacketsPerWindow = 100;
    internal const int MaxSources = 4_096;

    private readonly object _gate = new();
    private readonly Dictionary<IPAddress, SourceWindow> _sourceWindows = [];
    private long _globalWindowStartedMilliseconds;
    private int _globalPackets;

    internal int SourceCount
    {
        get
        {
            lock (_gate)
            {
                return _sourceWindows.Count;
            }
        }
    }

    internal bool TryAcquire(IPAddress? source, long nowMilliseconds)
    {
        lock (_gate)
        {
            if (nowMilliseconds - _globalWindowStartedMilliseconds >= WindowMilliseconds)
            {
                _globalWindowStartedMilliseconds = nowMilliseconds;
                _globalPackets = 0;
            }
            if (++_globalPackets > GlobalPacketsPerWindow || source is null)
            {
                return false;
            }

            source = Normalize(source);
            if (!_sourceWindows.TryGetValue(source, out var window))
            {
                EvictIfNeeded(nowMilliseconds);
                window = new SourceWindow(nowMilliseconds, 0, nowMilliseconds);
                _sourceWindows[source] = window;
            }
            if (nowMilliseconds - window.StartedMilliseconds >= WindowMilliseconds)
            {
                window.StartedMilliseconds = nowMilliseconds;
                window.Packets = 0;
            }
            window.LastSeenMilliseconds = nowMilliseconds;
            return ++window.Packets <= SourcePacketsPerWindow;
        }
    }

    internal void Cleanup(long nowMilliseconds)
    {
        lock (_gate)
        {
            foreach (var source in _sourceWindows
                         .Where(pair => nowMilliseconds - pair.Value.LastSeenMilliseconds > SourceTtlMilliseconds)
                         .Select(pair => pair.Key)
                         .ToList())
            {
                _sourceWindows.Remove(source);
            }
        }
    }

    private void EvictIfNeeded(long nowMilliseconds)
    {
        CleanupLocked(nowMilliseconds);
        while (_sourceWindows.Count >= MaxSources)
        {
            var oldest = _sourceWindows.MinBy(pair => pair.Value.LastSeenMilliseconds);
            _sourceWindows.Remove(oldest.Key);
        }
    }

    private void CleanupLocked(long nowMilliseconds)
    {
        foreach (var source in _sourceWindows
                     .Where(pair => nowMilliseconds - pair.Value.LastSeenMilliseconds > SourceTtlMilliseconds)
                     .Select(pair => pair.Key)
                     .ToList())
        {
            _sourceWindows.Remove(source);
        }
    }

    private static IPAddress Normalize(IPAddress source)
        => source.IsIPv4MappedToIPv6 ? source.MapToIPv4() : source;

    private sealed class SourceWindow(
        long startedMilliseconds,
        int packets,
        long lastSeenMilliseconds)
    {
        public long StartedMilliseconds { get; set; } = startedMilliseconds;

        public int Packets { get; set; } = packets;

        public long LastSeenMilliseconds { get; set; } = lastSeenMilliseconds;
    }
}
